### PR TITLE
[survey] Fix for JSONdata and LINST instruments

### LINKED ISSUE #7667: [survey_accounts] BMI Calculator missing questions / Cannot submit BMI survey
**Describe the bug**
The BMI Calculator survey is missing questions and therefore cannot be submitted. 

BMI Calculator on test-loris-dev loris ca

**To Reproduce**

1. Go to the survey accounts module and click a BMI calculator survey 

- 1. If you select "Metric" as the measurement, the error "Error running scoring algorithm" appears at the top left corner of the page.
BMI Calculator Error running scoring algorithm

- 2. If you select "Imperial" as the measurement, the error "Update statement did not execute successfully" appears 
BMI Calculator Update Statement did not execute successfully

**What did you expect to happen?**
Expected height and weight fields to appear. Here is what the BMI Calculator looks like on demo.loris.ca:

BMI Calculator on demo loris ca

**Browser Environment**
- Chrome 93.0.4577.82
- Firefox 92.0.1

### PR DESCRIPTION
**Pull request by ridz1208**
## Brief summary of changes
The survey instrument logic had fallen severely behind the latest standards. This PR adds support for LINST instrument pages, `getSubtestList()` which deprecated the `instrument_subtests` table in the database and JSON data saving and loading.

MAKE SURE TO REVIEW PR with "Hide whitespaces changes" as the indentation in files was fixed for readability

More details on changes:
1. fixes numbering issues on pages in survey instruments (causing next/prev page to not function properly)
2. fixes type hinting issues, more specifically on return values from `getPrevPage()` and `getNextPage()` functions
3. fixes saving/loading functionality to support JSON data in survey forms
4. cleans redundant page number variables and unused variables (ie. `$totalpages`)
5. fixes #7667 (hopefully)

#### Testing instructions (if applicable)

1. Follow survey testing plan
2. make sure to test for at least 1 of each of the following usecases
    - PHP instrument using SQL table for saving
    - LINST instrument using SQL table
    - PHP instrument using JSON saving in flag table
    - LINST instrument using JSON saving in flag table
    - single page instrument
    - multipage instrument

#### Link(s) to related issue(s)

* Resolves #7667

### CODE REVIEW

**Comment by ridz1208** [comment]
@driusan critical ?

File: php/libraries/NDB_BVL_Instrument.class.inc [hunk 1]
```diff
@@ -1403,11 +1404,7 @@ abstract class NDB_BVL_Instrument extends NDB_Page
     function getFieldValue(string $field)
     {
         $allValues = $this->getInstanceData();
-
-        if (!array_key_exists($field, $allValues)) {
-            throw new \OutOfBoundsException("Invalid field for instrument");
```
Comment (ridz1208) [inline php/libraries/NDB_BVL_Instrument.class.inc:1408]: @driusan might be necessary for 23 as well

**Comment by driusan** [comment]
@ridz1208 is going to split this up into multiple PRs. Parts of it (the NDB_BVL_Instrument part) are critical, parts of it (the survey part) can go into a bug fix release after being well tested, since the changes seem pretty significant.

**Comment by ridz1208** [comment]
needs rebase afte #7819 and #7820 are merged

**Comment by ridz1208** [comment]
blocked by #7818

**Comment by ridz1208** [comment]
@driusan I think this was deemed as not critical to release but it's still a valueable bugfix if we wanna assign it. It is rebased and ready to go

**Comment by ridz1208** [comment]
Hey @suzanne-lee given that you have the initial ticket for this. can you try out this fix ?

**Comment by suzanne-lee** [comment]
For BMI Calculator, the top left corner now says "Page 1 of 4" rather than "Page 1 of 1". However, the save & continue button is no longer appearing and therefore cannot be fully tested. See screenshot:

Screen Shot 2022-01-11 at 9 37 21 AM

**Comment by ridz1208** [comment]
@suzanne-lee should be ready again

**Comment by ridz1208** [comment]
@suzanne-lee I removed the print statement and fixed the page numbers shown.

As for your saving issue, I do not get that error when I test, here are a couple debugging steps you can take:
1. make sure your bmi.linst and bmi.meta files are freshly copied from the raisinbread directory
2. set the `jsonData` variable in `bmi.meta` to `true` and see if the pages save. (if they save then you might have an SQL issue in your bmi table)
3. set the `jsonData` back to `false` and see if the issue you get is related to the SQL table you have for the bmi
4. try other instruments that are non LINST


I tried all steps above and they all work on my VM. Also the error you reported seems to be linked to the getDataDictionary() function in the NDB_BVL_Instrument_LINST class. If all else fails, maybe you can try to print the return of that function and see why the Date_taken field is not in that array

**Comment by suzanne-lee** [comment]
Still trying to figure out the Date_taken problem, but the page numbering has been resolved.

**Comment by suzanne-lee** [comment]
I am able to view and submit all of the pages, as well as submit the entire form. However, the "Choose unit of measurement" question doesn't save; the element's name is supposed to be `unit_classification` but is `pageNum` on the browser (see below).

Screen Shot 2022-04-05 at 10 08 28 AM

**Comment by cmadjar** [comment]
@ridz1208 there are conflicts with the base branch. Also, should there be an entry in the CHANGELOG for this bug fix?

**Comment by ridz1208** [comment]
@driusan conflicts resolved

**Pull request merged**